Added support for Topology Manager to Device Manager

lmdaly opened the pull request:
**What this PR does / why we need it**:
Adding support for Topology Manager to Device Manager 
Topology Manager based merged design proposal here: kubernetes/community#1680
Issue for tracking PRs: #72828
**What type of PR is this?**

> /kind feature



**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

Commit:
Added support for Topology Manager to Device Manager

lmdaly commented:
/cc @ConnorDoyle @balajismaniam @nolancon

RenaudWasTaken commented:
/ok-to-test

RenaudWasTaken commented on pkg/kubelet/cm/devicemanager/endpoint.go:
```diff
@@ -87,6 +91,14 @@ func (e *endpointImpl) callback(resourceName string, devices []pluginapi.Device)
 	e.cb(resourceName, devices)
 }
 
+func (e *endpointImpl) getDevices(resourceName string) []pluginapi.Device {
```
Could you use the m.healthyDevices map in the manager rather than creating this map here?

Having the endpoint maintain its own map of Devices was what we used to do in the past but it created a lot of complexity.

Currently, the manager maintains its own map of Devices through populating `m.healthyDevices` and `m.unhealthyDevices` could we rewrite these maps in a way that would work for this use case?

One idea could be to change the type of these maps to pluginapi.Device and generate the sets in the GetCapacity function.

What do you think?

lmdaly replied:
We can't use healthyDevices as it as because it doesn't contain the topology information. What you suggested is totally plausible, or if we could add an additional map to the manager for allDevices that is a map[string][]pluginapi.Device. This could be populated at the same time as healthy/unhealthy devices in the callback.

Pretty small addition that leaves the healthy/unhealthy devices untouched,

RenaudWasTaken commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -138,6 +145,117 @@ func newManagerImpl(socketPath string) (*ManagerImpl, error) {
 }
 
+//GetTopologyHints calculates a BitMask based on device socket availability for an incoming container
+func (m *ManagerImpl) GetTopologyHints(pod v1.Pod, container v1.Container) topologymanager.TopologyHints {
```
It might be good to move this function to its own file and split it in multiple sub functions to help with readability

lmdaly replied:
Sure, I'll do that!

RenaudWasTaken reviewed (commented):
You're missing some tests that would help increase confidence in these code paths :)

spiffxp commented:
/hold
ref: https://github.com/kubernetes/enhancements/issues/693#issuecomment-466728227

lmdaly commented:
> You're missing some tests that would help increase confidence in these code paths :)

We were planning on adding the tests in a separate PR. Would it be easier to review if the test was in this pr?

Sreemanti-Ghosh committed:
Unit test for devicemanager/topology.go

resouer commented on pkg/kubelet/cm/devicemanager/topology.go:
```diff
@@ -0,0 +1,205 @@
+)
+
+type topology struct {
+	largestSocket int64
```
Having a device specific attributes like `largestSocket` and `deviceMask` in a generic topology struct in Device Manager seems problematic a bit. This makes the whole design read more like "NUMA Topology awareness in node" instead of "Topology awareness in node", which does not quite align with description in https://github.com/kubernetes/enhancements/issues/693 unless I missed sth. /cc @vishh 

Per WG discussion we believe we are very glad to see topology awareness in node start from NUMA while I just expect a more generic struct and implementation which could apply to NUMA, GPU and other up-coming  topology aware devices etc.  /cc @RenaudWasTaken

resouer commented on pkg/kubelet/cm/devicemanager/topology.go:
```diff
@@ -0,0 +1,205 @@
+}
+
+// GetTopologyHints calculates a BitMask based on device socket availability for an incoming container
+func (m *ManagerImpl) GetTopologyHints(pod v1.Pod, container v1.Container) topologymanager.TopologyHints {
```
One example is, I do think this part of implementation could be decoupled from `topology.go` as specific impl for NUMA.

In detail, could we have sth like: `pkg/kubelet/cm/devicemanager/{tolology.go, numa.go, nv_link.go}`?

lmdaly replied:
@resouer would it be possible to get a diagram/description on how the nv link would work with device plugins & planned topology manager?

With NV Link, the idea is that two devices have an affinity is my understanding whereas the current proposal deals only with affinity between cpu and devices on a socket level. 

With an understanding of how NV Link could work with this proposal I'll have a better idea on how to abstract the current implementation to allow for such an extension in the future,

lmdaly replied:
@resouer  can you look at the updates to this PR? The socketmask implementation has changed and so the topology has changed

Commit:
Make TopologyHints more intuitive - device manager (#28)

Commit:
Update Device Manager to align with SocketMask Interface

*Update Topology file
*Update Device Allocation logic

klueska commented on pkg/kubelet/cm/devicemanager/topology.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+func checkIfMaskEqualsStoreMask(existingDeviceHints []topologymanager.TopologyHint, newMask socketmask.SocketMask) bool {
+	maskEqual := false
```
No need for this variable, just do the following below:
```
for _, storedHint := range existingDeviceHints {
		if storedHint.SocketMask.IsEqual(newMask) {
			return true
		}
	}
	return false
```

klueska commented on pkg/kubelet/cm/devicemanager/topology.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+func calculateIfDeviceHasSocketAffinity(deviceHints []topologymanager.TopologyHint) bool {
+	admit := false
```
No need for this variable, just do the following below:
```
for _, hint := range deviceHints {
	if hint.SocketMask.Count() == 1 {
		return true
	}
}
return false
```

kad commented:
Code in this PR has several places where `nil` from plugins that don't expose TopologyInfo will lead to crashes similar to below:
```
I0530 22:13:31.509435   23834 topology_manager.go:161] [topologymanager] Topology Admit Handler
I0530 22:13:31.509477   23834 topology_manager.go:164] [topologymanager] Pod QoS Level: BestEffort
I0530 22:13:31.509517   23834 topology_manager.go:181] [topologymanager] Topology Manager only affinitises Guaranteed pods.
I0530 22:13:31.509743   23834 manager.go:631] Topology Affinities for pod 034dd8d9-830f-11e9-bb76-b8aeed7c7427 container intelgpu-de
mo-job-1 are: {<nil>}
E0530 22:13:31.510307   23834 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/work/go/go/src/runtime/panic.go:522
/work/go/go/src/runtime/panic.go:82
/work/go/go/src/runtime/signal_unix.go:390
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager.go:634
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager.go:693
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager.go:335
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/container_manager_linux.go:654
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/lifecycle/predicate.go:72
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1768
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2051
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1889
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1832
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1436
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:1053
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153
/work/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/work/go/go/src/runtime/asm_amd64.s:1337
```

ConnorDoyle commented:
/kind feature

Commit:
Update Device Manager to use new Topology Manager algorithm
Update Device Manager to take Device Plugins with a nil Topology
into account
Move Device Manager creation with Topology Manager from PR#74357 to
remove dependency

Commit:
Update Device Manager Tests to match Topology Manager Implementation
Update Build Files

kad commented on pkg/kubelet/cm/devicemanager/topology.go:
```diff
@@ -0,0 +1,126 @@
+			topologyAligned = true
+		}
+	}
+	return topologyAligned
```
1. no need in `topologyAligned` variable. 
```go
for _, device := range allDevices[resource] {
	topology := device.Topology
	if topology != nil {
		return true
	}
}
return false
```
2. beside checking nil, most probably it is necessary to check that socket mask is not zero (at least one bit is set).

lmdaly replied:
Thanks will remove the variable.

This is not a socket mask here, just an int. I will check for a -1

Commit:
Minor edits to accomodate new topology manager semantics

cbf123 commented on pkg/kubelet/cm/devicemanager/topology.go:
```diff
@@ -72,11 +73,9 @@ func (m *ManagerImpl) GetTopologyHints(pod v1.Pod, container v1.Container) ([]to
+		klog.Infof("[devicemanager-topology] DeviceHints: %v", deviceHints)
 	}
 
 	return deviceHints, admit
```
I think you're returning two values here but the new function signature at line 13 only specifies one return value.

cbf123 commented on pkg/kubelet/cm/devicemanager/manager_stub.go:
```diff
@@ -65,8 +65,8 @@ func (h *ManagerStub) GetWatcherHandler() cache.PluginHandler {
-func (h *ManagerStub) GetTopologyHints(pod v1.Pod, container v1.Container) ([]topologymanager.TopologyHint, bool) {
-	return []topologymanager.TopologyHints{}, false
+func (h *ManagerStub) GetTopologyHints(pod v1.Pod, container v1.Container) []topologymanager.TopologyHint {
+	return []topologymanager.TopologyHints{}
```
should that be "return []topologymanager.TopologyHint{}" without the "s" at the end?

cbf123 commented on pkg/kubelet/cm/devicemanager/topology.go:
```diff
@@ -10,8 +10,7 @@ import (
+func (m *ManagerImpl) GetTopologyHints(pod v1.Pod, container v1.Container) []topologymanager.TopologyHint {
 
 	var deviceHints []topologymanager.TopologyHint
 	admit := false
```
I think "admit := false" should be removed if we're not returning it.

cbf123 commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -649,21 +649,32 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
+	availableTopologyAligned := available
+
 	sockets := make(map[int]bool)
 	socketsArray := containerTopologyHint.SocketMask.GetSockets()
```
I think this should be SocketAffinity, not SocketMask

lmdaly replied:
Added a condition around this section for TopologyHint being nil

cbf123 commented on pkg/kubelet/cm/devicemanager/topology.go:
```diff
@@ -107,23 +117,10 @@ func getDevicesPerSocket(resource string, available sets.String, allDevices map[
 func checkIfMaskEqualsStoreMask(existingDeviceHints []topologymanager.TopologyHint, newMask socketmask.SocketMask) bool {
-	maskEqual := false
 	for _, storedHint := range existingDeviceHints {
 		if storedHint.SocketMask.IsEqual(newMask) {
```
I think this should be SocketAffinity instead of SocketMask

klueska commented:
@cbf123 This PR hasn't been updated since it was pulled out of the 1.15 release. The developers have updated it locally, but not pushed it yet. It's probably not worth doing an extensive review until they've pushed the latest.

cbf123 commented:
@klueska  Thanks for the heads-up...I was trying to pull in the patches to try them out, so the ones I commented on are the ones that actually caused problems.

Do you know offhand whether anyone has proposed any examples of plugin-side changes yet?

klueska commented:
@cbf123 Im not 100% sure. We (Nvidia) have been reviewing the patches and have adopted the core TopologyManager and CPUManager changes into our internal stack, but have our own patches for the devicemanager that don’t require any plugin changes. Our custom changes aren’t generic enough to be upstreamable (they hard code GPU awareness into the devicemanager) but we wanted to start using as much of this stack as possible while it stabilizes.

cbf123 commented on pkg/kubelet/cm/devicemanager/topology.go:
```diff
@@ -0,0 +1,126 @@
+
+func checkIfDeviceHasSocketAlignment(resource string, allDevices map[string][]pluginapi.Device) bool {
+	topologyAligned := false
+	for _, device := range allDevices[resource] {
```
the for loop here seems odd.  If we assume that the devices for a given resource all either have or don't have a topology then we should only need to look at the first one.  If they might not have a topology, then shouldn't we return "false" if *any* of them don't have a topology?

lmdaly replied:
Changed this and assumed that all or none have topology associated. Thanks!

cbf123 commented on pkg/kubelet/cm/devicemanager/topology.go:
```diff
@@ -26,34 +25,36 @@ func (m *ManagerImpl) GetTopologyHints(pod v1.Pod, container v1.Container) ([]to
 
 			for socket, amountAvail := range deviceSocketAvail {
-				klog.Infof("Socket: %v, Avail: %v AmountReq: %v", socket, amountAvail, amount)
 				mask, _ := socketmask.NewSocketMask(int(socket))
```
I think this could go inside the conditional at line 51 since we don't need it if that conditional evaluates to "false"

lmdaly commented:
> Do you know offhand whether anyone has proposed any examples of plugin-side changes yet?

@cbf123 there's two dummy device plugins that we have been using for testing purposes here - https://github.com/lmdaly/kubernetes/tree/dev/topology_manager_new_algorithm/cluster/addons/device-plugins used in conjunction with PR: https://github.com/kubernetes/kubernetes/pull/74423

Also thanks for the review I'll get to the comments today/tomorrow.

Commit:
Merge pull request #36 from nolancon/device-manager-edits

Minor edits to accomodate new topology manager semantics

cbf123 commented:
@lmdaly Thanks.  I see that you co-authored sriov-network-device-plugin originally.  Do you know where the equivalent changes would need to be made in that package?

cbf123 commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -653,7 +653,7 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
 
 	sockets := make(map[int]bool)
-	socketsArray := containerTopologyHint.SocketMask.GetSockets()
+	socketsArray := containerTopologyHint.SocketAffinity.GetSockets()
```
As I said in my earlier comment, I think that containerTopologyHint can be nil if the feature gate is disabled.  I think we need a conditional around this whole section of code.

klueska commented:
I had a private conversation with @lmdaly about this a few weeks ago, but I'm curious what other people think.

I’m starting to work through the case where I don’t just want to allocate GPUs, but also NICs via the `devicemanager`. The hints for GPU devices should be independent of the hints for NICs, in terms of how the `TopologyManager` *ands* them together into a merged hint.

Right now the `devicemanager` does a sort of "pre-merge" of hints from all device types before passing this "pre-merged" hint back to the `TopologyManager`. It does this by looking at all the devices requested by a container and creating a single set of hints where both devices are available.

For example, a NIC on either numa node and a GPU on numa node 0 would result in a hint returned by the `devicemanager` of {{1,0}{1,1}}.

It seems strange to me to have two levels of merging however -- once done by the `devicemanager` and then again done by the `TopologyManager.`

I would expect the implementation `devicemanager`’s `GetTopologyHints()` call to look something like:
```
hints map[string][]TopologyHint
foreach request {
  if isDeviceResource(request.resource) {
      hints[request.resource] := plugin[request.resource].GetTopologyHints()
  }
}
return hints
```

Then leave it completely up to the `TopologyManager` to merge them.

From the perspective of the implementor of the GPU device plugin and the NIC device plugin, I don’t want to have to think about two different merge strategies going on above me when I return my hints. I’d rather just know that I’m going to be *anded* with all other resource types that provide hints, and then provide my hints accordingly.

Moreover, it seems strange to me to treat "device resources" differently from other resources on the system (e.g. CPUs) in terms of how they provide their hints. Each device type should have an equal opportunity to provide hints directly to the `TopologyManager` the same way the CPUManager does and have them merged together using a single, well-defined strategy.

Providing these semantics, however, would require an API change in the `GetTopologyHints()` call:
```
from: GetTopologyHints(pod v1.Pod, container v1.Container) []TopologyHint
to:   GetTopologyHints(pod v1.Pod, container v1.Container) map[string][]TopologyHint
```

Where the new map being returned is indexed by the resource name.

With this in place, the `CPUManager` can continue doing things as it currently is, but return:
```
return map[string][]int{
    "cpu": []TopologyHint{},
}

instead of:

return []TopologyHint{}
```

And the `devicemanager` would return hints as indicated above:
```
hints map[string][]TopologyHint
foreach request {
  if isDeviceResource(request.resource) {
      hints[request.resource] := plugin[request.resource].GetTopologyHints()
  }
}
return hints
```

What do people think?

cbf123 commented:
In what circumstances would this actually matter to the implementor of the device plugin?  When would it make a difference whether there is one merge strategy or more?

If we were allowing the user to specify something like "I want strict affinity for resource A, but I can tolerate non-affinity for resource B" then I could see it making a more significant difference.

klueska commented:
But this is exactly the type of thing I could imagine a future TopologyManager policy doing.

Right now the policies are all or nothing:
1)`strict` implies you will only have the pod admitted if there is a common `preferred` NUMA affinity (where `preferred` is defined by the field in https://github.com/kubernetes/kubernetes/pull/78516/files#diff-a19f204f56e91b0871491b88c0e0b139L51)
2) `preferred` implies the pod will always be admitted, but if there is no common `preferred` NUMA affinity, the returned hint will be "all-numa-nodes"

A future policy could say:
3) `best-effort` implies the pod will always be admitted, but if there is no common `preferred` NUMA affinity, we will choose the affinity that *most* resources preferred

cbf123 commented:
I'm not against the idea of passing all the hints back to the TopologyManager if we think it'll allow us to make better decisions.
If we think we might want to get fancier later though, it might make sense to allow the specification of strict/preferred to be done at the pod level rather than be specified at the node level.
I'm not sure I see a useful reason to distinguish between preferred and best-effort, what you describe as best-effort seems like the logical way to implement "preferred" if we have the information to do it.

klueska commented:
I'm actually only just now looking at the code changes here in detail. It looks like there is actually a misalignment between what I expected these changes to do and what they actually do (i.e. we took a different strategy for our internal `devicemanager` changes built on top of the `TopologyManager`).

I expected these changes to give each device plugin the chance to supply its own set of `TopologyHints` each time a container was being launched and a specific number of devices were being requested. 

it actually doesn't do this at all though. Instead, a new field called `Topology` has been added to the `Device` abstraction that a device plugin simply sets when it reports a device to the `devicemanager`. When it comes time to determine the `TopologyHints` from the `devicemanager` it simply consults this new `Topology` field for all devices. If it finds a common alignment among all registered device types, it sets an appropriate hint.

Is there a reason not to delegate the hint generation out to the device plugin? It seems they are in the best position to return the proper set of hints given (1) the number of devices being requested, and (2) the current set of devices that have already been allocated.

I know for sure this is the case with Nvidia GPUs, as we rely heavily on considering NVLINKs as well as each GPUs position in the PCIe hierarchy in order to determine the best set of hints to return for a given GPU request (again this is in our custom stack but we'd like to migrate to fully upstream eventually).

klueska commented:
The canonical example being the one I mentioned in my feedback document here:
https://docs.google.com/document/d/1UbD1igeBA1sRuqwkQgJ5sJ9sf1rao_LLORWLhwSdGUo/edit#heading=h.im70k4qvof1

As far as I can see, there is no way to apply the logic below in a device independent manner without delegating the hint generation out to the Nvidia GPU device plugin. 

```
For example, 8 GPUs, 4 per socket, all available:

Request of 1 GPU would give:
    []TopologyHint{
        TopologyHint{
            SocketMask: 10,
            Preferred: true,
        },
        TopologyHint{
            SocketMask: 01,
            Preferred: true,
        },
    
Request of 2-4 GPUs would give:
    []TopologyHint{
        TopologyHint{
            SocketMask: 10,
            Preferred: true,
        },
        TopologyHint{
            SocketMask: 01,
            Preferred: true,
        },
        TopologyHint{
            SocketMask: 11,
            Preferred: false,
        },
    
Request of 5-8 GPUs would give:
    []TopologyHint{
        TopologyHint{
            SocketMask: 11,
            Preferred: true,
        },
    
With 8 GPUs, 4 per socket, 3 GPUs per socket already allocated:

Request of 1 GPU would give:
    []TopologyHint{
        TopologyHint{
            SocketMask: 01,
            Preferred: true,
        },
        TopologyHint{
            SocketMask: 10,
            Preferred: true,
        },

Request of 2 GPUs would give:
    []TopologyHint{
        TopologyHint{
            SocketMask: 11,
            Preferred: false,
        },
```

cbf123 commented:
The example you give only depends on the number of devices on each NUMA node, so wouldn't devicemanager have that information?

I think your mention of NVLINK is more useful, as it represents a more complicated topology than just "which GPU is on which NUMA node".

klueska commented:
Yes, I guess a better example would have been:

```
4 GPUs, 2 GPUs per socket,
GPUs on socket 0 have NVLINK,
GPUs on socket 1 do not:

Request of 1 GPU would give:
    []TopologyHint{
        TopologyHint{
            SocketMask: 01,
            Preferred: true,
        },
        TopologyHint{
            SocketMask: 10,
            Preferred: true,
        },

Request of 2 GPUs would give:
    []TopologyHint{
        TopologyHint{
            SocketMask: 01,
            Preferred: true,
        },
    []TopologyHint{
        TopologyHint{
            SocketMask: 10,
            Preferred: false,
        },
    []TopologyHint{
        TopologyHint{
            SocketMask: 11,
            Preferred: false,
        },

Request of 3-4 GPUs would give:
    []TopologyHint{
        TopologyHint{
            SocketMask: 11,
            Preferred: true,
        }
```

lmdaly commented:
> @lmdaly Thanks. I see that you co-authored sriov-network-device-plugin originally. Do you know where the equivalent changes would need to be made in that package?

I would imagine during device discovery ie. of the VFs there would be an additional call to discover the Socket ID of the device via the /sys/class/net filesystem.

moshe010 commented on pkg/kubelet/cm/devicemanager/topology.go:
```diff
@@ -0,0 +1,124 @@
+				allDeviceSocketsInt = append(allDeviceSocketsInt, socket)
+			}
+			crossSocketMask, _ := socketmask.NewSocketMask(allDeviceSocketsInt...)
+			deviceHints = append(deviceHints, topologymanager.TopologyHint{SocketAffinity: crossSocketMask, Preferred: false})
```
It seem that for scenario when you request resources which have to spread across NUMA nodes the topology hint will be Preferred= false.   
example: If a pod requests 4 VFs and I have 2 VFs on NUMA 0 and 2 VF on NUMA 1 free. The topology hint will be (1,1  Preferred=false).  I would expect that the Preferred will be set according "preferred := len(deviceHints) == 0".
So if we don't have deviceHints per NUMA then the cross NUMA should be Preferred=True.

lmdaly replied:
Yes you're correct that should be preferred true in that case. Need to update this PR so that will be included, thanks!

moshe010 commented on pkg/kubelet/cm/devicemanager/topology.go:
```diff
@@ -0,0 +1,124 @@
+package devicemanager
```
so in the cpumanager you have the toplogy_hints.go, so maybe we should rename this file to topology_hints.go as well (for consistency)

lmdaly replied:
+1 good suggestion will update

moshe010 commented:
> @lmdaly Thanks. I see that you co-authored sriov-network-device-plugin originally. Do you know where the equivalent changes would need to be made in that package?
@cbf123 if you want sriov-network-device-plugin changes you can take them from here [1]. I did them for my internal testing.
[1] https://github.com/moshe010/sriov-network-device-plugin/tree/numa

Commit:
Updates to device manager based on comments
Renamed toplology.go to topology_hints.go, applied same to test file
Added logic for a true preferred in the case the number of total devices
on a socket is less than requested number

klueska commented on pkg/kubelet/cm/devicemanager/topology_hints_test.go:
```diff
@@ -0,0 +1,91 @@
+
+		testContainer.Resources.Requests = testResourceList
+		hints := m.GetTopologyHints(testPod, testContainer)
+		if reflect.DeepEqual(hints, tc.expectedHints) {
```
should be:
```
- if reflect.DeepEqual(hints, tc.expectedHints) {
+ if !reflect.DeepEqual(hints, tc.expectedHints) {
```

klueska replied:
Actually, looking at this entire test suite more closely, these tests are not complete, which is why they are now failing.

I have updated the tests in my follow-on branch to this patch, where I modify `GetTopologyHints()` to return a map instead of a slice.
https://github.com/klueska/kubernetes/commit/d6bebfd7c84dad51241898a8d752ea452cfbbc0a

If we are OK leaving the tests broken until my PR based on this branch is merged, then we shouldn't have to change anything here yet.

klueska commented on pkg/kubelet/cm/devicemanager/topology_hints_test.go:
```diff
@@ -0,0 +1,91 @@
+
+		testDevices := make(map[string][]pluginapi.Device)
+		testDevice1 := pluginapi.Device{ID: "ID", Health: pluginapi.Healthy, Topology: &pluginapi.TopologyInfo{Socket: int64(0)}}
+		testDevice2 := pluginapi.Device{ID: "ID", Health: pluginapi.Healthy, Topology: &pluginapi.TopologyInfo{Socket: int64(1)}}
```
This will need to change with the new change made in https://github.com/kubernetes/kubernetes/pull/74423

```
testDevice1 := pluginapi.Device{ID: "ID", Health: pluginapi.Healthy, Topology: &pluginapi.TopologyInfo{Node: &pluginapi.NUMANode{0}}}
testDevice2 := pluginapi.Device{ID: "ID", Health: pluginapi.Healthy, Topology: &pluginapi.TopologyInfo{Node: &pluginapi.NUMANode{1}}}
```

lmdaly replied:
Yep I need to update this PR to reflect those changes, I just wanted to wait until we got a consensus on #74423 before making those changes

klueska commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -288,9 +291,10 @@ func cleanup(t *testing.T, m Manager, p *Stub) {
+	testManager, err := newManagerImpl(socketName, topologyStore)
 	as := assert.New(t)
 	as.NotNil(testManager)
 	as.Nil(err)
```
This patch also needs to include the following change in order for the original `devicemanager` tests to pass:
```
@@ -602,6 +606,7 @@ func getTestManager(tmpDir string, activePods ActivePodsFunc, testRes []TestReso
        allocatedDevices:  make(map[string]sets.String),
        endpoints:         make(map[string]endpointInfo),
        podDevices:        make(podDevices),
+       topologyAffinityStore: topologymanager.NewFakeManager(),
        activePods:        activePods,
        sourcesReady:      &sourcesReadyStub{},
        checkpointManager: ckm,
```

lmdaly commented:
Closing this PR as it is superseded by https://github.com/kubernetes/kubernetes/pull/80570 
(This PR would have had to be basically rewritten after the PR https://github.com/kubernetes/kubernetes/pull/80569 merged)

klueska commented:
@moshe010 with this PR closed and #80569 and #80570 in place. You should be able to base your PR proposing the `Merge()`abstraction on top of them.

moshe010 commented:
@klueska , thank I will start working is next week. I also created https://github.com/kubernetes/enhancements/pull/1160 on the Merge()abstraction.
